### PR TITLE
CI: Make tests more generic

### PR DESCRIPTION
**Pull request by anomiex**
#### Changes proposed in this Pull Request:

The general idea here is that each project should define composer commands for the different types of tests we support, and the infrastructure should just run them without doing too much that's specific to any one plugin.

Thus, the PHP, JS, and Coverage workflows are rewritten to move all the Jetpack-specific stuff into Jetpack, and as a side effect the packages also get some changes to be more flexible than they had been previously.

I also added some logic to better avoid running unnecessary tests: A change that only touches the Beta plugin, for example, probably doesn't need Jetpack's tests run. But a change that touches the Autoloader package will also need Jetpack tests to ensure that it didn't break Jetpack.

I punted for now on the E2E tests. They're done in their own weirdly complicated way so it'll take more work to figure out which parts can be made generic, which are Jetpack-specific, and which if any we might be able to get rid of entirely.

#### Jetpack product discussion

1199714989172614-as-1199667028950985

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Do tests pass? Note many of them are renamed.
* Does the coverage get uploaded correctly to both coverage sites?

#### Proposed changelog entry for your changes:

* None needed.

### CODE REVIEW

File: .github/files/list-changed-projects.php [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+		}
+		ksort( $cache );
+	}
+
+	return $cache;
+}
+
+/**
+ * Fetch the list of changed projects.
+ * `
```
Comment (kraftbj) [inline .github/files/list-changed-projects.php:65]: This is a lonely little tick.

**Comment by anomiex** [comment]
I had a shower thought this morning, I could probably combine the three test-*.yml workflows into one to improve the DRY there. They're already nearly identical.

File: .github/files/list-changed-projects.php [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+/**
+ * Fetch the list of changed projects.
+ *
+ * @return string[]
+ */
+function get_changed_projects() {
+	global $infrastructure_files, $ignore_globs;
+
+	// Code coverage probably needs to run all tests to avoid confusing metrics.
+	// @todo Determine if that's actually true.
```
Comment (anomiex) [inline .github/files/list-changed-projects.php:77]: Looks like codecov.io can handle [carrying forward from past runs](https://docs.codecov.io/docs/carryforward-flags), but codeclimate.com can't.

OTOH, we might be able to use something like [this action](https://github.com/marketplace/actions/download-base-artifact) to try to download a coverage artifact from a past run and copy it into the coverage directory, if we really want.

File: .github/files/list-changed-projects.php [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	'.github/files/setup-wordpress-env.sh',
+	'.github/php-version',
+	'.github/workflows/tests.yml',
+	'.nvmrc',
+);
+
+// Globs that should be ignored. Supports `*`, `**`, and `?`.
+// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+$ignore_globs = array(
+	'**.md',
```
Comment (kraftbj) [inline .github/files/list-changed-projects.php:26]:
```suggestion
	'**.md',
	'**.txt',
```

for plugin readme.txt files, at least.

File: .github/files/list-changed-projects.php [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		sprintf( 'git diff --no-renames --name-only %s...%s', escapeshellarg( $base ), escapeshellarg( $head ) ),
+		array( array( 'pipe', 'r' ), array( 'pipe', 'w' ), STDERR ),
+		$pipes
+	);
+	if ( ! $p ) {
+		exit( 1 );
+	}
+	fclose( $pipes[0] );
+
+	// PHP doesn't seem to have something that'll correctly handle `**`, so munge to a regex.
```
Comment (kraftbj) [inline .github/files/list-changed-projects.php:117]: We should add some generic unit tests for monorepo tools. This is one of those things that I would feel better having an unit test to confirm it is handling the `**` as we'd expect.

Could also make a handy small package of itself or general handy php utils.

File: composer.json [hunk 5]
```diff
@@ -61,6 +61,9 @@
 			"printf '\\e[37;41m`composer php:requirelist` is deprecated. Use `composer phpcs:lint:required` and/or `composer phpcs:fix:required` instead.\\e[0m\\n'",
 			"@phpcs:fix:required",
 			"@phpcs:lint:required"
+		],
+		"test-js": [
+			"cd tools/cli && yarn install && yarn test"
```
Comment (kraftbj) [inline composer.json:66]: ❤️

**Review by kraftbj — CHANGES_REQUESTED** [review]

**Review by kraftbj — CHANGES_REQUESTED** [review]
needs a quick master merge, but looks good.

File: .github/files/list-changed-projects.php (hunk 4 above)
Comment (anomiex) [reply .github/files/list-changed-projects.php:118]: Tests probably would be useful. We can add such tests easily enough now by defining or adding to the appropriate scripts in the root-level `composer.json`, someone would just have to pick a framework and such and then write them.

As its own package, that'd be cool but I'd want to add support for more advanced globs and such.

**Review by kraftbj — APPROVED** [review]

**Comment by anomiex** [comment]
r220630-wpcom

**Pull request merged**